src/Sylius/Bundle/CoreBundle/Provider/SalesPerPeriodProvider.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Provider;
+
+use Doctrine\ORM\EntityRepository;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\OrderPaymentStates;
+use Sylius\Component\Core\Sales\Mapper\SalesPeriodMapperInterface;
+use Sylius\Component\Core\Sales\Provider\SalesPerPeriodProviderInterface;
+use Sylius\Component\Core\Sales\ValueObject\SalesPeriod;
+
+final class SalesPerPeriodProvider implements SalesPerPeriodProviderInterface
+{
+    /** @param EntityRepository<OrderInterface> $orderRepository */
+    public function __construct(private EntityRepository $orderRepository, private SalesPeriodMapperInterface $salesPeriodMapper)
+    {
+    }
+
+    public function provide(SalesPeriod $salesPeriod, ChannelInterface $channel): array
+    {
+        $queryBuilder = $this->orderRepository->createQueryBuilder('o')
+            ->select('SUM(o.total) AS total')
+            ->andWhere('o.paymentState = :state')
+            ->andWhere('o.channel = :channel')
+            ->setParameter('state', OrderPaymentStates::STATE_PAID)
+            ->setParameter('channel', $channel)
+        ;
+
+        switch ($salesPeriod->getInterval()) {
+            case 'year':
+                $queryBuilder
+                    ->addSelect('YEAR(o.checkoutCompletedAt) as year')
+                    ->groupBy('year')
+                    ->andWhere('YEAR(o.checkoutCompletedAt) >= :startYear AND YEAR(o.checkoutCompletedAt) <= :endYear')
+                    ->setParameter('startYear', $salesPeriod->getStartDate()->format('Y'))
+                    ->setParameter('endYear', $salesPeriod->getEndDate()->format('Y'))
+                ;
+                $dateFormatter = static function (\DateTimeInterface $date): string {
+                    return $date->format('Y');
+                };
+                $resultFormatter = static function (array $data): string {
+                    return (string) $data['year'];
+                };
+
+                break;
```

Comment:
Before the `switch` statement:
```php
$interval = $salesPeriod->getInterval();
```

```suggestion
            case 'year':
                $this->modifyQueryBuilderForHandlingYear($queryBuilder)
                break;
```

And after the `switch` statement

```php
$dateFormatter = $this->getDateFormatterFor($interval);
$resultFormatter = $this->getResultFormatterFor($interval);
```

It'd improve readability IMO.

Reply:
`modifyQueryBuilderForHandlingYear` can be replaced by any other name. Just an example, nothing better came to my mind.